poor diagnostic due to overloading in `extern "C"` block

https://godbolt.org/z/Koo18frzd (GCC does a bit of a better job, treating them as redeclarations and mentioning they're C functions)

The following invalid C++ code:
```
extern "C" {
void Foo(int);
int Foo();
}
```
will give the error:
```
<source>:5:5: error: functions that differ only in their return type cannot be overloaded
    5 | int Foo();
      | ~~~ ^
<source>:4:6: note: previous declaration is here
    4 | void Foo(int);
      | ~~~~ ^
1 error generated.
```
which is profoundly unfair to the person writing the code. The error isn't anything to do with functions that differ only in their return type, and everything to do with being in an extern "C" block, where overloading does not exist.

(quoting @ssbr from an internal bug)

Confirmed.

The diagnostic is here:

https://github.com/llvm/llvm-project/blob/0c361270afff83cd6433cf865ed5a410dadfb33f/clang/lib/Sema/SemaDecl.cpp#L3945-L3948

if we modified the `if` statement:

https://github.com/llvm/llvm-project/blob/0c361270afff83cd6433cf865ed5a410dadfb33f/clang/lib/Sema/SemaDecl.cpp#L3921

to

```cpp
if (getLangOpts().CPlusPlus && !(New->isExternC() && Old->isExternC())) {
```

The diagnostic is:

```console
externC.cpp:3:5: error: conflicting types for 'Foo'
    3 | int Foo();
      |     ^
externC.cpp:2:6: note: previous declaration is here
    2 | void Foo(int);
      |      ^
1 error generated.
```

Which is a lot closer to gcc.

CC @AaronBallman

Thanks for the confirmation. I will give it a try.
I didn't know there was an `overloadable` attribute in clang, so I'll check it out as I go along.

In gcc, the compile error message for the program is as follows:
```
test.cpp:3:13: error: conflicting declaration of C function 'int Foo()'
    3 |         int Foo();
      |             ^~~
test.cpp:2:14: note: previous declaration 'void Foo(int)'
    2 |         void Foo(int);
      |              ^~~
```

This message explicitly states that there is a declaration conflict in C, but there seems to be no such error message in clang.
In order to output the same error message as gcc, I think we need to implement a new error message and branching process, is the above suggested modification to output 
`externC.cpp:3:5: error: conflicting types for 'Foo'`
sufficient?

I would appreciate any advice anyone can give me.

--- a/clang/docs/ReleaseNotes.rst
+++ b/clang/docs/ReleaseNotes.rst
@@ -261,7 +261,7 @@ Improvements to Clang's diagnostics
   compilation speed with modules. This warning is disabled by default and it needs
   to be explicitly enabled or by ``-Weverything``.
 
-- Improved diagnostics when trying to overload a function in extern "C". (#GH80235)
+- Improved diagnostic when trying to overload a function in an ``extern "C"`` context. (#GH80235)
 
 Improvements to Clang's time-trace
 ----------------------------------

--- a/clang/docs/ReleaseNotes.rst
+++ b/clang/docs/ReleaseNotes.rst
@@ -261,6 +261,8 @@ Improvements to Clang's diagnostics
   compilation speed with modules. This warning is disabled by default and it needs
   to be explicitly enabled or by ``-Weverything``.
 
+- Improved diagnostics when trying to overload a function in extern "C". (#GH80235)
+
 Improvements to Clang's time-trace
 ----------------------------------
 

--- a/clang/lib/Sema/SemaDecl.cpp
+++ b/clang/lib/Sema/SemaDecl.cpp
@@ -3882,6 +3882,10 @@ bool Sema::MergeFunctionDecl(FunctionDecl *New, NamedDecl *&OldD, Scope *S,
         if (New->isCXXClassMember() && New->isOutOfLine())
           Diag(New->getLocation(), diag::err_member_def_does_not_match_ret_type)
               << New << New->getReturnTypeSourceRange();
+        else if (Old->isExternC() && New->isExternC() &&
+                 !Old->hasAttr<OverloadableAttr>() &&
+                 !New->hasAttr<OverloadableAttr>())
+          Diag(New->getLocation(), diag::err_conflicting_types) << New;
         else
           Diag(New->getLocation(), diag::err_ovl_diff_return_type)
               << New->getReturnTypeSourceRange();

--- a/clang/test/SemaCXX/extern-c.cpp
+++ b/clang/test/SemaCXX/extern-c.cpp
@@ -77,6 +77,19 @@ namespace foo {
   }
 }
 
+namespace extern_ovl {
+  extern "C" {
+    __attribute__((overloadable))
+    void ovl_decl(void); // expected-note {{previous}}
+
+    __attribute__((overloadable))
+    int ovl_decl(int);
+
+    __attribute__((overloadable))
+    int ovl_decl(void); // expected-error {{functions that differ only in their return type}}
+  }
+}
+
 namespace linkage {
   namespace redecl {
     extern "C" {
@@ -86,6 +99,12 @@ namespace linkage {
       void linkage_redecl(int); // ok, still not extern "C"
       void linkage_redecl(float); // expected-note {{previous}}
       void linkage_redecl(double); // expected-error {{conflicting types}}
+    }
+  }
+  namespace redecl_2 {
+    extern "C" {
+      void linkage_redecl_2(); // expected-note {{previous}}
+      int linkage_redecl_2(int); // expected-error {{conflicting types}}
     }
   }
   namespace from_outer {

